### ISSUE TITLE
Running as a user with no home directory

### ISSUE THREAD

**Issue by bentley** [micahflee/onionshare#943]
On OpenBSD, we often provide [daemon control scripts](https://man.openbsd.org/rc.d.8) for packages that users install, so the daemon can be started automatically on bootup and flags can be saved in config files. Typically the daemon runs as a dedicated user created by our package manager with a home directory of `/var/empty`.

I tried something consistent with how we usually set up our packages: running as an unprivileged user `_onionshare`, executing `onionshare --config /etc/onionshare.json --receive`, and a `_onionshare`‐owned `data_dir` of `/var/onionshare`. Unfortunately, it fails because OnionShare always tries to create `~/.config`:

    Traceback (most recent call last):
      File "/usr/local/bin/onionshare", line 22, in <module>
        onionshare.main()
      File "/usr/local/lib/python3.6/site-packages/onionshare/__init__.py", line 41, in main
        common.load_settings()
      File "/usr/local/lib/python3.6/site-packages/onionshare/common.py", line 55, in load_settings
        self.settings = Settings(self, config)
      File "/usr/local/lib/python3.6/site-packages/onionshare/settings.py", line 48, in __init__
        self.filename = self.build_filename()
      File "/usr/local/lib/python3.6/site-packages/onionshare/settings.py", line 139, in build_filename
        return os.path.join(self.common.build_data_dir(), 'onionshare.json')
      File "/usr/local/lib/python3.6/site-packages/onionshare/common.py", line 143, in build_data_dir
        os.makedirs(onionshare_data_dir, 0o700, True)
      File "/usr/local/lib/python3.6/os.py", line 210, in makedirs
        makedirs(head, mode, exist_ok)
      File "/usr/local/lib/python3.6/os.py", line 220, in makedirs
        mkdir(name, mode)
    PermissionError: [Errno 13] Permission denied: '/var/empty/.config'

It would be nice if `onionshare_data_dir` could be set through a flag or a JSON option.

**Comment by mig5** [comment]
This would hopefully be fully solved by #366 , but indeed we should not attempt to create the .config directory if we have already passed a custom settings json file.

**Comment by mig5** [comment]
Long story but after some investigation, this regression actually relates to the localisation of CLI strings. We have to load settings 'early' and this has inadvertently resulted in triggering logic that tries the default settings json location.

Noting it for my own reference just because #879 will probably unblock this.

**Comment by mig5** [comment]
@micahflee this might be fixed due to https://github.com/micahflee/onionshare/pull/1336 ?

**Comment by micahflee** [comment]
Oh quite possibly. What's $XDG_CONFIG_HOME set to when there's no home directory? What's the easiest way to test if this is fixed?

**Comment by mig5** [comment]
> What's $XDG_CONFIG_HOME set to when there's no home directory?

I think it's an empty variable with no value... it seems it's actually surprisingly hard to create a user with no home dir.. if I `usermod -d '' someuser`, I can't even `su -s /bin/bash someuser` because of the lack of home dir. That's on Linux though.

> What's the easiest way to test if this is fixed?

I created a test user `onionshare` in my VM with a home dir of `/var/empty`

Then with `XDG_CONFIG_HOME` set to `/var/onionshare` (which I created in advance with ownership for user `onionshare`) I was able to run OnionShare just fine, the data ends up in `/var/onionshare/`.

```
XDG_CONFIG_HOME=/var/onionshare onionshare-cli /tmp/onionshare/README.md --verbose
```

```
root@onionshare:/home/user# ls -al /var/onionshare/onionshare/tmp/tmppr98yh0r/
total 7956
drwx------ 3 onionshare nogroup    4096 May  4 08:17 .
drwx------ 3 onionshare nogroup    4096 May  4 08:16 ..
-rw------- 1 onionshare nogroup   20472 May  4 08:17 cached-certs
-rw------- 1 onionshare nogroup 2291600 May  4 08:17 cached-microdesc-consensus
-rw------- 1 onionshare nogroup 5804281 May  4 08:17 cached-microdescs.new
srw------- 1 onionshare nogroup       0 May  4 08:16 control_socket
-rw------- 1 onionshare nogroup      32 May  4 08:16 cookie
drwx------ 2 onionshare nogroup    4096 May  4 08:16 keys
-rw------- 1 onionshare nogroup       0 May  4 08:16 lock
-rw------- 1 onionshare nogroup     212 May  4 08:16 state
-rw-r--r-- 1 onionshare nogroup     334 May  4 08:16 torrc
```

So it's hard to test on Linux without a valid home dir to begin with, but I think the matter stands that if you set `XDG_CONFIG_HOME` to any arbitrary directory, so long as it exists or the onionshare user can create it, the 'data' dir will be created in that location rather than the hardcoded `~/.config` .

**Comment by mig5** [comment]
I feel like this can be closed thanks to #1336 . You can use `$XDG_CONFIG_HOME` to get OnionShare to store its config/state data in some other location than the home directory. But please reopen if it continues to present problems for BSD packaging.
